crates/turborepo-lib/src/run/watch.rs

```diff
@@ -1,37 +1,231 @@
+                let event = event.unwrap();
+                Self::handle_change_event(
+                    &mut run,
+                    event.event.unwrap(),
+                    &mut current_runs,
+                    &base,
+                    &telemetry,
+                    &handler,
+                )
+                .await?;
+            }
+
+            Ok::<(), Error>(())
+        };
+
+        select! {
+            _ = event_fut => {}
+            _ = subscriber.listen() => {
+                tracing::info!("shutting down");
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_change_event(
+        run: &mut Run,
+        event: proto::package_change_event::Event,
+        current_runs: &mut HashMap<PackageName, JoinHandle<Result<i32, run::Error>>>,
+        base: &CommandBase,
+        telemetry: &CommandEventBuilder,
+        handler: &SignalHandler,
+    ) -> Result<(), Error> {
+        // Should we recover here?
+        match event {
+            proto::package_change_event::Event::PackageChanged(proto::PackageChanged {
+                package_name,
+            }) => {
+                let package_name = PackageName::from(package_name);
+                // If not in the filtered pkgs, ignore
+                if !run.filtered_pkgs.contains(&package_name) {
+                    return Ok(());
                 }
-                proto::package_change_event::Event::RediscoverPackages(_) => {
-                    println!("Rediscovering packages");
+
+                let mut args = base.args().clone();
+                args.command = args.command.map(|c| {
+                    if let Command::Watch(execution_args) = c {
+                        Command::Run {
+                            execution_args: Box::new(ExecutionArgs {
+                                filter: vec![format!("...{}", package_name)],
+                                ..*execution_args
+                            }),
+                            run_args: Box::new(RunArgs {
+                                no_cache: true,
+                                daemon: true,
+                                ..Default::default()
+                            }),
+                        }
+                    } else {
+                        unreachable!()
+                    }
+                });
+
+                let new_base =
+                    CommandBase::new(args, base.repo_root.clone(), get_version(), base.ui.clone());
+
+                // TODO: Add logic on when to abort vs wait
+                if let Some(run) = current_runs.remove(&package_name) {
+                    run.abort();
                 }
+
+                let telemetry = telemetry.clone();
+                let handler = handler.clone();
+                current_runs.insert(
+                    package_name,
+                    tokio::spawn(async move {
+                        let mut run = RunBuilder::new(new_base)?
```

Comment:
Perhaps now is the time for an alternate constructor that is not from `CommandBase`?

Reply:
Maybe as a follow up